### PR TITLE
Make History metrics available to other callbacks on epoch end

### PR DESCRIPTION
**Pull request by gw0**
When chaining callbacks it would make sense to also have an average `loss` metric available in `on_epoch_end()` (not just `val_loss`). With this modification callbacks that monitor a value become useful even if you do not provide any validation data (and without writing a custom Callback class).

```
callbacks = [
    ModelCheckpoint(monitor='loss', mode='min', filepath='foo.hdf5', save_best_only=True),
    EarlyStopping(monitor='loss', mode='min', patience=100),
]
```

### CODE REVIEW

**Comment by fchollet** [comment]
See: https://github.com/fchollet/keras/pull/1696

**Comment by fchollet** [comment]
This is fixed by moving metrics averages computation to `BaseLogger`, and moving `History` and `ProgbarLogger` to the end of the callbacks list.
